ReactSkia/sdk/NopollWebsocket.cpp

```diff
@@ -0,0 +1,298 @@
+/*
+* Copyright (C) 1994-2022 OpenTV, Inc. and Nagravision S.A.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+#include <nopoll.h>
+
+#include "ReactSkia/utils/RnsLog.h"
+#include "NopollWebsocket.h"
+
+using namespace std;
+namespace facebook {
+namespace react {
+
+std::queue<NopollRequest*> RequestQueue;
+int currentWorkerThreadCount = 0;
+std::mutex threadCountLock;
+std::mutex requestQueueLock;
+sem_t startDownloadingThread;  /*semaphore to signal downloading thread to start, once there is something in the list*/
+
+NopollWebsocket* NopollWebsocket::sharedNopollWebsocket_{nullptr};
+std::mutex NopollWebsocket::requestLock_;
+
+NopollWebsocket::NopollWebsocket() {
+  ctx_ = nopoll_ctx_new ();
+  sem_init(&startDownloadingThread, 0, 0);
+  wsMessageThread_ = std::thread([this]() {
+    std::thread t1(&NopollWebsocket::pollThread,this);
+    int error_code;
+    while (ctx_) {
+      // wait for ever
+      error_code = nopoll_loop_wait (ctx_, 0);
+      if (error_code == -4)
+        RNS_LOG_ERROR(" io waiting mechanism, errno=" << errno);
+      // recover by just calling io wait engine
+      // try to limit recoveries to avoid infinite loop
+    }
+  });
+}
+
+NopollWebsocket* NopollWebsocket::sharedNopollWebsocket() {
+  std::lock_guard<std::mutex> lock(requestLock_);
+  if(sharedNopollWebsocket_ == nullptr)
+    sharedNopollWebsocket_ = new NopollWebsocket();
+  return sharedNopollWebsocket_;
+}
+
+NopollWebsocket::~NopollWebsocket() {
+  nopoll_ctx_unref(ctx_);
+  std::lock_guard<std::mutex> lock(requestLock_);
+  if(this == sharedNopollWebsocket_)
+    sharedNopollWebsocket_ = nullptr;
+}
+
+void NopollWebsocket::multiConn(){
+  NopollRequest* nopollRequest = nullptr;
+  while(true) { // forever
+    requestQueueLock.lock();
+    if(RequestQueue.size() !=0) {
+      nopollRequest =RequestQueue.front();
+      RequestQueue.pop();
+    } else {
+      threadCountLock.lock();
+      currentWorkerThreadCount--;
+      threadCountLock.unlock();
+      requestQueueLock.unlock();
+      break;
+    }
+    requestQueueLock.unlock();
+    if(!nopollRequest)
+      continue;
```

Comment:
I think better to move this check after we call front() on queue

Reply:
moved.